api/src/infrastructure/kubernetes/infrastructure.rs

```diff
@@ -0,0 +1,894 @@
+    fn from(err: KubeError) -> Self {
+        KubernetesInfrastructureError::UnexpectedError {
+            internal_message: err.to_string(),
+        }
+    }
+}
+
+impl From<ServiceError> for KubernetesInfrastructureError {
+    fn from(err: ServiceError) -> Self {
+        match err {
+            ServiceError::InvalidServiceType { label } => {
+                KubernetesInfrastructureError::UnknownServiceType {
+                    unknown_label: label,
+                }
+            }
+            err => KubernetesInfrastructureError::UnexpectedError {
+                internal_message: err.to_string(),
+            },
+        }
+    }
+}
+
+impl From<ServiceBuilderError> for KubernetesInfrastructureError {
+    fn from(err: ServiceBuilderError) -> Self {
+        KubernetesInfrastructureError::UnexpectedError {
+            internal_message: err.to_string(),
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use kube::api::{ObjectMeta, TypeMeta};
+    use std::collections::BTreeMap;
+
+    macro_rules! deployment_object {
+        ($deployment_name:expr, $app_name:expr, $service_name:expr, $container_type:expr) => {{
+            let mut labels = BTreeMap::new();
+
+            if let Some(app_name) = $app_name {
+                labels.insert(String::from(APP_NAME_LABEL), app_name);
+            }
+            if let Some(service_name) = $service_name {
+                labels.insert(String::from(SERVICE_NAME_LABEL), service_name);
+            }
+            if let Some(container_type) = $container_type {
+                labels.insert(String::from(CONTAINER_TYPE_LABEL), container_type);
+            }
+
+            Object {
+                types: TypeMeta::default(),
+                metadata: ObjectMeta {
+                    name: String::from($deployment_name),
+                    namespace: None,
+                    labels,
+                    annotations: BTreeMap::new(),
+                    resourceVersion: None,
+                    ownerReferences: vec![],
+                    uid: None,
+                    generation: None,
+                    generateName: None,
+                    initializers: None,
+                    finalizers: vec![],
+                    creation_timestamp: None,
+                    deletion_timestamp: None,
+                },
+                spec: DeploymentSpec::default(),
+                status: None,
+            }
+        }};
+    }
+
+    #[test]
+    fn should_parse_service_from_deployment_spec() {
+        let deployment = deployment_object!(
+            "master-nginx",
+            Some(String::from("master")),
+            Some(String::from("nginx")),
+            None
+        );
+
+        let service = ServiceBuilder::try_from(deployment)
+            .unwrap()
+            .id("same-random-id".to_string())
+            .started_at(Utc::now())
+            .build()
+            .unwrap();
+
+        // TODO: deployment name generation
```

Comment:
Is this a missing feature, or a missing test?

Reply:
`id("same-random-id".to_string())` had to be removed. Nice catch.